// @jeefo/material/components/appbar/index.js
 jeefo.register("node_modules/@jeefo/material/components/appbar/index.js", async (exports, module) => { const __dirname = "node_modules/@jeefo/material/components/appbar", __filename = "node_modules/@jeefo/material/components/appbar/index.js"; const require = path => { return jeefo.require(path, __filename, __dirname); }; /* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
* File Name   : app_bar.js
* Created at  : 2021-01-14
* Updated at  : 2021-01-21
* Author      : jeefo
* Purpose     :
* Description :
* Reference   :
.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.*/
// ignore:start
"use strict";

/* globals*/
/* exported*/

// ignore:end

const TranscludeController = (await require("@jeefo/component/transclude_controller"));

const transcluder = new TranscludeController((`<md-surface color="{{$md_app_bar.color}}"><jf-content></jf-content></md-surface>`
/* space filler


*/));

exports.style = (`.md-app-bar{display:block}`
/* space filler


*/);

exports.template = element =>  {
    element.setAttribute("md-elevation", 4);
    element.classList.add("md-app-bar");
    transcluder.transclude(element);
};

exports.bindings = {
    color : '@',
};

exports.controller = (function () { function MDAppBar () {} 
    MDAppBar.prototype.on_init = function () {
    };
MDAppBar.__jeefo_class__ = true; return MDAppBar;}());

exports.controller_name = "$md_app_bar";
 });
