// @jeefo/component/definitions_table.js
 jeefo.register("node_modules/@jeefo/component/definitions_table.js", async (exports, module) => { const __dirname = "node_modules/@jeefo/component", __filename = "node_modules/@jeefo/component/definitions_table.js"; const require = path => { return jeefo.require(path, __filename, __dirname); }; /* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
* File Name   : definitions_table.js
* Created at  : 2019-06-24
* Updated at  : 2019-09-10
* Author      : jeefo
* Purpose     :
* Description :
* Reference   :
.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.*/
// ignore:start
"use strict";

/* globals*/
/* exported*/

// ignore:end

const ComponentDefinition = (await require("./component_definition"));
const DirectiveDefinition = (await require("./directive_definition"));

var DefinitionsTable = (function () { 
    function DefinitionsTable () {
        this.components = Object.create(null);
        this.directives = Object.create(null);
    }

    DefinitionsTable.prototype.register_component = function (selectors, path) {
        if (typeof selectors === "string") {
            selectors = [selectors];
        }
        const definition = new ComponentDefinition(selectors, path);
        selectors.forEach(name => {
            if (this.components[name]) {
                throw new Error("Duplicated component selector registered");
            }
            this.components[name] = definition;
        });
    };

    DefinitionsTable.prototype.register_directive = function (selectors, path) {
        if (typeof selectors === "string") {
            selectors = [selectors];
        }
        const definition = new DirectiveDefinition(selectors, path);
        selectors.forEach(name => {
            if (this.directives[name]) {
                throw new Error("Duplicated component selector registered");
            }
            this.directives[name] = definition;
        });
    };

    DefinitionsTable.prototype.register = function (selectors, path) {
        if (typeof selectors === "string") {
            selectors = [selectors];
        }
        if (! path.endsWith(".js")) {
            path += ".js";
        }

        if (path.endsWith("_component.js")) {
            this.register_component(selectors, path);
        } else if (this.path.endsWith("_directive.js")) {
            this.register_directive(selectors, path);
        } else {
            throw new Error("Invalid path name");
        }
    };

    DefinitionsTable.prototype.get_component = async function (name) {
        const component_definition = this.components[name];
        if (component_definition) {
            if (! component_definition.is_resolved) {
                await component_definition.resolve();
            }

            return component_definition;
        }
    };

    DefinitionsTable.prototype.get_directive = async function (name) {
        const directive_definition = this.directives[name];
        if (directive_definition) {
            if (! directive_definition.is_resolved) {
                await directive_definition.resolve();
            }

            return directive_definition;
        }
    };
DefinitionsTable.__jeefo_class__ = true; return DefinitionsTable;}());

module.exports = new DefinitionsTable();
 });
