// @jeefo/component/directive_definition.js
 jeefo.register("node_modules/@jeefo/component/directive_definition.js", async (exports, module) => { const __dirname = "node_modules/@jeefo/component", __filename = "node_modules/@jeefo/component/directive_definition.js"; const require = path => { return jeefo.require(path, __filename, __dirname); }; /* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
* File Name   : directive_definition.js
* Created at  : 2017-08-07
* Updated at  : 2021-01-06
* Author      : jeefo
* Purpose     :
* Description :
* Reference   :
.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.*/
// ignore:start
"use strict";

/* globals jeefo*/
/* exported*/

// ignore:end

const for_each      = (await require("@jeefo/utils/object/for_each"));
const extend_member = (await require("@jeefo/utils/class/extend_member"));
const styles        = (await require("./styles"));
const IDefinition   = (await require("./interfaces/i_definition"));

const is_class = value => (
    value.toString().startsWith("class") ||
    value.__jeefo_class__
);

var DirectiveDefinition = (function () {  
    function DirectiveDefinition (selectors, path) {
        IDefinition.call(this, selectors, path);
        this.priority = 0;
    } DirectiveDefinition.prototype = Object.create(IDefinition.prototype);

    DirectiveDefinition.prototype.resolve = async function () {
        const {
            type, priority, style, controller, controller_name,
            bindings, dependencies = {}
        } = await jeefo.require(this.path);

        if (priority) { this.priority = priority; }
        if (type) {
            if (type.toLowerCase() === "structure") {
                this.is_structure = true;
            } else {
                throw new SyntaxError("Invalid directive type");
            }
        }

        if (style) {
            const selectors = this.selectors.map(s => `"${s}"`);
            styles.add_style(style, {
                "directive-selectors" : `[${ selectors.join(", ") }]`
            });
        }

        // Conroller
        if (controller) {
            let Ctrl;
            if (typeof controller === "function") {
                if (is_class(controller)) {
                    Ctrl = controller;
                } else {
                    function Controller () {}
                    extend_member(Controller, "on_init", controller);
                    Ctrl = Controller;
                }
            } else {
                function Controller () {}
                for_each(controller, (key, value) => {
                    extend_member(Controller, key, value);
                });
                Ctrl = Controller;
            }
            this.Controller = Ctrl;
            if (controller_name) {
                this.controller_name = controller_name;
            }
        }

        IDefinition.prototype.set_binders.call(this, bindings);
        IDefinition.prototype.set_dependencies.call(this, dependencies);
        this.is_resolved = true;
    };
DirectiveDefinition.__jeefo_class__ = true; return DirectiveDefinition;}());

module.exports = DirectiveDefinition;
 });
