// @jeefo/component/text_binder.js
 jeefo.register("node_modules/@jeefo/component/text_binder.js", async (exports, module) => { const __dirname = "node_modules/@jeefo/component", __filename = "node_modules/@jeefo/component/text_binder.js"; const require = path => { return jeefo.require(path, __filename, __dirname); }; /* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
* File Name   : text_binder.js
* Created at  : 2021-02-18
* Updated at  : 2021-02-19
* Author      : jeefo
* Purpose     :
* Description :
* Reference   :
.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.*/
// ignore:start
"use strict";

/* globals*/
/* exported*/

// ignore:end

const Interpreter = (await require("./interpreter"));

var TextBinder = (function () { 
    function TextBinder (node, component) {
        this.node            = node;
        this.bound_component = component;
    }

    TextBinder.prototype.initialize = function () {
        const script = `\`${this.node.textContent}\``;
        this.interpreter = new Interpreter(script, this.bound_component);
        this.node.textContent = this.value = this.interpreter.get_value();
    };

    TextBinder.prototype.digest = function () {
        const value = this.interpreter.get_value();
        if (this.value !== value) {
            this.node.textContent = value;
            this.value = value;
        }
    };
TextBinder.__jeefo_class__ = true; return TextBinder;}());

module.exports = TextBinder;
 });
