// @jeefo/component/component_definition.js
 jeefo.register("node_modules/@jeefo/component/component_definition.js", async (exports, module) => { const __dirname = "node_modules/@jeefo/component", __filename = "node_modules/@jeefo/component/component_definition.js"; const require = path => { return jeefo.require(path, __filename, __dirname); }; /* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
* File Name   : component_definition.js
* Created at  : 2019-06-24
* Updated at  : 2021-01-07
* Author      : jeefo
* Purpose     :
* Description :
* Reference   :
.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.*/
// ignore:start
"use strict";

/* globals jeefo*/
/* exported*/

// ignore:end

const extend_member        = (await require("@jeefo/utils/class/extend_member"));
const object_for_each      = (await require("@jeefo/utils/object/for_each"));
const styles               = (await require("./styles"));
const IDefinition          = (await require("./interfaces/i_definition"));
const TranscludeController = (await require("./transclude_controller"));

const is_class = v => v.__jeefo_class__ || v.toString().startsWith("class");

const pendings = Object.create(null);

const _resolve = async instance => {
    const {
        type, style, template, controller, controller_name,
        bindings, dependencies = {}
    } = await jeefo.require(instance.path);

    // Type
    if (type) {
        if (type.toLowerCase() === "structure") {
            instance.is_structure = true;
        } else {
            throw new SyntaxError("Invalid definition type");
        }
    }

    // Style
    if (style) {
        const selectors = instance.selectors.map(s => `"${s.toLowerCase()}"`);
        styles.add_style(style, {
            "component-selectors" : `[${ selectors.join(", ") }]`
        });
    }

    // Template
    if (typeof template === "string") {
        instance.transclude_controller = new TranscludeController(template);
    } else if (typeof template === "function") {
        instance.template_handler = template;
    } else {
        instance.transclude_controller = new TranscludeController();
    }

    // Conroller
    if (controller) {
        let Ctrl;
        if (typeof controller === "function") {
            if (is_class(controller)) {
                Ctrl = controller;
            } else {
                function Controller () {}
                extend_member(Controller, "on_init", controller);
                Ctrl = Controller;
            }
        } else {
            function Controller () {}
            object_for_each(controller, (key, value) => {
                extend_member(Controller, key, value);
            });
            Ctrl = Controller;
        }
        instance.Controller = Ctrl;
        if (controller_name) {
            instance.controller_name = controller_name;
        }
    }

    return {bindings, dependencies};
};

var ComponentDefinition = (function () {  
    function ComponentDefinition (selectors, path) {
        IDefinition.call(this, selectors, path);
        this.template_handler      = null;
        this.transclude_controller = null;
    } ComponentDefinition.prototype = Object.create(IDefinition.prototype);

    ComponentDefinition.prototype.resolve = function () {
        if (pendings[this.selectors[0]]) {
            return pendings[this.selectors[0]];
        }

        pendings[this.selectors[0]] = new Promise(async resolver => {
            const {bindings, dependencies} = await _resolve(this);
            IDefinition.prototype.set_binders.call(this, bindings);
            IDefinition.prototype.set_dependencies.call(this, dependencies);
            this.is_resolved = true;
            resolver();
        });

        return pendings[this.selectors[0]];
    };

    ComponentDefinition.prototype.transclude = function (element) {
        this.transclude_controller.transclude(element);
    };
ComponentDefinition.__jeefo_class__ = true; return ComponentDefinition;}());

module.exports = ComponentDefinition;
 });
