// @jeefo/component/index.js
 jeefo.register("node_modules/@jeefo/component/index.js", async (exports, module) => { const __dirname = "node_modules/@jeefo/component", __filename = "node_modules/@jeefo/component/index.js"; const require = path => { return jeefo.require(path, __filename, __dirname); }; /* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
* File Name   : index.js
* Created at  : 2017-08-08
* Updated at  : 2020-10-23
* Author      : jeefo
* Purpose     :
* Description :
* Reference   :
.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.*/
// ignore:start
"use strict";

/* globals*/
/* exported*/

// ignore:end

const definitions_table = (await require("./definitions_table"));

const REGEX_UNDERLINE = /_/g;

[
    "if",
    "for_each",
    "switch",
    "jf_bind",
    "jf_class",
    "jf_bind_html",
].forEach(filename => {
    const path     = `${__dirname}/directives/${filename}`;
    const selector = filename.replace(REGEX_UNDERLINE, '-');
    definitions_table.register_directive(selector, path);
});

const switch_placeholder_path = `${__dirname}/directives/switch_placeholder`;
definitions_table.register_component(
    "switch-placeholder", switch_placeholder_path
);

module.exports = {
    compile             : (await require("./compiler")),
    IComponent          : (await require("./interfaces/i_component")),
    IRenderable         : (await require("./interfaces/i_renderable")),
    IDefinition         : (await require("./interfaces/i_definition")),
    Directive           : (await require("./components/directive")),
    InvisibleComponent  : (await require("./components/invisible_component")),
    StructureComponent  : (await require("./components/structure_component")),
    RenderableComponent : (await require("./components/renderable_component")),

    definitions_table
};
 });
