// @jeefo/ecma_parser/es8/literals/boolean_literal.js
 jeefo.register("node_modules/@jeefo/ecma_parser/es8/literals/boolean_literal.js", async (exports, module) => { const __dirname = "node_modules/@jeefo/ecma_parser/es8/literals", __filename = "node_modules/@jeefo/ecma_parser/es8/literals/boolean_literal.js"; const require = path => { return jeefo.require(path, __filename, __dirname); }; /* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
* File Name   : boolean_literal.js
* Created at  : 2019-09-04
* Updated at  : 2020-08-23
* Author      : jeefo
* Purpose     :
* Description :
* Reference   :
.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.*/
// ignore:start
"use strict";

/* globals*/
/* exported*/

// ignore:end

const { EXPRESSION }      = (await require("../enums/precedence_enum"));
const { expression }      = (await require("../enums/states_enum.js"));
const { get_pre_comment } = (await require("../../helpers"));

module.exports = {
    id         : "Boolean literal",
    type       : "Literal",
    precedence : EXPRESSION,

    is         : (_, { current_state: s }) => s === expression,
    initialize : (node, token, parser) => {
        node.pre_comment = get_pre_comment(parser);
        node.value       = token.value;
        node.start       = token.start;
        node.end         = token.end;
    },

    protos : {
        is_valid_simple_assignment_target : () => false,
    }
};
 });
