// @jeefo/material/components/sidenav/index.js
 jeefo.register("node_modules/@jeefo/material/components/sidenav/index.js", async (exports, module) => { const __dirname = "node_modules/@jeefo/material/components/sidenav", __filename = "node_modules/@jeefo/material/components/sidenav/index.js"; const require = path => { return jeefo.require(path, __filename, __dirname); }; /* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
* File Name   : index.js
* Created at  : 2019-06-21
* Updated at  : 2021-02-23
* Author      : jeefo
* Purpose     :
* Description :
* Reference   :
.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.*/
// ignore:start
"use strict";

/* globals*/
/* exported*/

// ignore:end

const Observer = (await require("@jeefo/observer"));
const backdrop = (await require("../../utils/backdrop"));

const root = "md-sidenav";

var MDSidenavOptions = (function () { 
    var __JEEFO_ӨМНӨХ_ХЭСЭГ__ = function (MDSidenavOptions) { (function () { var __JEEFO_PROPERTY_НЭР__ = "variant"; var __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__ = Object.getOwnPropertyDescriptor(MDSidenavOptions.prototype, __JEEFO_PROPERTY_НЭР__) || { configurable: true, enumerable: false }; __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__.get = function      () { return this.instance.variant  || "side"; }; Object.defineProperty(MDSidenavOptions.prototype, __JEEFO_PROPERTY_НЭР__, __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__); }());
    (function () { var __JEEFO_PROPERTY_НЭР__ = "position"; var __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__ = Object.getOwnPropertyDescriptor(MDSidenavOptions.prototype, __JEEFO_PROPERTY_НЭР__) || { configurable: true, enumerable: false }; __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__.get = function     () { return this.instance.position || "left";   }; Object.defineProperty(MDSidenavOptions.prototype, __JEEFO_PROPERTY_НЭР__, __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__); }());
    (function () { var __JEEFO_PROPERTY_НЭР__ = "has_backdrop"; var __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__ = Object.getOwnPropertyDescriptor(MDSidenavOptions.prototype, __JEEFO_PROPERTY_НЭР__) || { configurable: true, enumerable: false }; __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__.get = function () { return this.instance.has_backdrop; }; Object.defineProperty(MDSidenavOptions.prototype, __JEEFO_PROPERTY_НЭР__, __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__); }());

    (function () { var __JEEFO_PROPERTY_НЭР__ = "is_open"; var __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__ = Object.getOwnPropertyDescriptor(MDSidenavOptions.prototype, __JEEFO_PROPERTY_НЭР__) || { configurable: true, enumerable: false }; __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__.get = function ()      { return this.instance.is_open;  }; Object.defineProperty(MDSidenavOptions.prototype, __JEEFO_PROPERTY_НЭР__, __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__); }());
    (function () { var __JEEFO_PROPERTY_НЭР__ = "is_open"; var __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__ = Object.getOwnPropertyDescriptor(MDSidenavOptions.prototype, __JEEFO_PROPERTY_НЭР__) || { configurable: true, enumerable: false }; __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__.set = function (value) { this.instance.is_open = value; }; Object.defineProperty(MDSidenavOptions.prototype, __JEEFO_PROPERTY_НЭР__, __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__); }()); };

    function MDSidenavOptions (instance, $element) {
        this.$element = $element;
        this.instance = instance;
    } __JEEFO_ӨМНӨХ_ХЭСЭГ__(MDSidenavOptions);
MDSidenavOptions.__jeefo_class__ = true; return MDSidenavOptions;}());

var BaseController = (function () { 
    function BaseController (options) {
        this.options = options;
        this._change_position(options.position);
    }

    BaseController.prototype._change_position = function (value) {
        const {$element} = this.options;
        switch (value) {
            case "left":
                $element.remove_class(`${root}--right`);
                $element.add_class(`${root}--left`);
                break;
            case "right":
                $element.remove_class(`${root}--left`);
                $element.add_class(`${root}--right`);
                break;
        }
    };
BaseController.__jeefo_class__ = true; return BaseController;}());

var MDSidenavSideController = (function () {  
    function MDSidenavSideController (options) {
        BaseController.call(this, options);
        this.class_name = `${root}--side`;

        const {$element} = options;
        const $wrapper = this.$wrapper = $element.parent().parent();
        const $content = this.$content = $wrapper.first(
            ".md-sidenav-container__content"
        );

        const {width} = $element;
        $wrapper.style("width", `calc(100% + ${width}px)`);
        $element.add_class(this.class_name);

        switch (options.position.toLowerCase()) {
            case "left" :
                $element.remove_class(this.side_class_name);
                this.side_class_name = `${root}--left`;
                $element.add_class(this.side_class_name);
                $content.style("paddingLeft", `${width}px`);
                $wrapper.style("marginLeft", `-${width}px`);
                break;
            case "right" :
                $element.remove_class(this.side_class_name);
                this.side_class_name = `${root}--right`;
                $element.add_class(this.side_class_name);
                $content.style("paddingRight", `${width}px`);
                $wrapper.style("marginRight", `-${width}px`);
                break;
        }

        if (options.is_open) this.open();
        $wrapper.style("transition", "none");
        $wrapper.trigger_reflow();
        $wrapper.style("transition", null);
    } MDSidenavSideController.prototype = Object.create(BaseController.prototype);

    MDSidenavSideController.prototype.open = function () {
        const {$wrapper, options: {$element, position}} = this;

        switch (position.toLowerCase()) {
            case "left" :
                $wrapper.style("paddingLeft", `${$element.width}px`);
                break;
            case "right" :
                $wrapper.style("paddingRight", `${$element.width}px`);
                break;
        }
    };

    MDSidenavSideController.prototype.close = function () {
        const {$wrapper, options: {position}} = this;

        switch (position.toLowerCase()) {
            case "left" :
                $wrapper.style("paddingLeft", null);
                break;
            case "right" :
                $wrapper.style("paddingRight", null);
                break;
        }
    };

    MDSidenavSideController.prototype.change_position = function (value) {
        const {$wrapper, $content, options} = this;
        const {$element, position, is_open} = options;
        const width = `${$element.width}px`;
        this._change_position(value);

        $wrapper.style("transition", "none");
        switch (position.toLowerCase()) {
            case "left" :
                $wrapper.css({
                    marginLeft   : `-${width}`,
                    paddingLeft  : is_open ? width : null,
                    marginRight  : null,
                    paddingRight : null,
                });
                $content.css({
                    paddingLeft  : width,
                    paddingRight : null,
                });
                break;
            case "right" :
                $wrapper.css({
                    marginLeft   : null,
                    paddingLeft  : null,
                    marginRight  : `-${width}`,
                    paddingRight : is_open ? width : null,
                });
                $content.css({
                    paddingLeft  : null,
                    paddingRight : width,
                });
                break;
        }
        $wrapper.trigger_reflow();
        $wrapper.style("transition", null);
    };

    MDSidenavSideController.prototype.destroy = function () {
        const {$wrapper, $content, options: {$element, position}} = this;

        switch (position.toLowerCase()) {
            case "left" :
                $wrapper.css({
                    marginLeft  : null,
                    paddingLeft : null,
                });
                $content.style("paddingLeft", null);
                break;
            case "right" :
                $wrapper.css({
                    marginRight  : null,
                    paddingRight : null,
                });
                $content.style("paddingRight", null);
                break;
        }
        $element.remove_class(this.class_name);
        $wrapper.css({
            width      : null,
            transition : "none",
        });
        $wrapper.trigger_reflow();
        $wrapper.style("transition", null);
    };
MDSidenavSideController.__jeefo_class__ = true; return MDSidenavSideController;}());

var MDSidenavOverController = (function () {  
    function MDSidenavOverController (options) {
        BaseController.call(this, options);
        const {$element} = options;

        this.class_name = `${root}--over`;
        $element.add_class(this.class_name);
        this[options.is_open ? "open" : "close"]();

        $element.style("transition", "none");
        $element.trigger_reflow();
        $element.style("transition", null);
    } MDSidenavOverController.prototype = Object.create(BaseController.prototype);

    MDSidenavOverController.prototype.open = function () {
        const {$element, has_backdrop} = this.options;
        $element.style("transform", `translateX(0)`);

        if (has_backdrop) {
            this.$backdrop = backdrop();
            this.$backdrop.on("click", () => this.close());
            $element.before(this.$backdrop);
        }
    };

    MDSidenavOverController.prototype.close = function () {
        const {$element, position} = this.options;
        switch (position.toLowerCase()) {
            case "left" :
                $element.style("transform", `translateX(-100%)`);
                break;
            case "right" :
                $element.style("transform", `translateX(100%)`);
                break;
        }

        if (this.$backdrop) {
            this.$backdrop.remove();
            this.$backdrop = null;
        }

        this.options.is_open = false;
    };

    MDSidenavOverController.prototype.change_position = function (value) {
        const {$element, position, is_open} = this.options;
        this._change_position(value);

        $element.style("transition", "none");
        if (! is_open) {
            switch (position.toLowerCase()) {
                case "left" :
                    $element.style("transform", `translateX(-100%)`);
                    break;
                case "right" :
                    $element.style("transform", `translateX(100%)`);
                    break;
            }
        }
        $element.trigger_reflow();
        $element.style("transition", null);
    };

    MDSidenavOverController.prototype.destroy = function () {
        const {$element} = this.options;
        $element.style("transform", null);
        $element.remove_class(this.class_name);
        if (this.$backdrop) {
            this.$backdrop.remove();
            this.$backdrop = null;
        }
    };
MDSidenavOverController.__jeefo_class__ = true; return MDSidenavOverController;}());

exports.selector = "md-sidenav";

exports.bindings = {
    variant       : "@",
    is_open       : "=isOpen",
    position      : "@",
    has_backdrop  : "<hasBackdrop",
    backdropClass : "@",
};

exports.template = `<md-surface><jf-content></jf-content></md-surface>`;

exports.controller = (function () { function MDSidenav () {} 
    MDSidenav.prototype.on_init = function ($element) {
        let controller;
        const options  = new MDSidenavOptions(this, $element);
        const observer = new Observer(this);
        $element.add_class("md-sidenav");

        if ([null, void 0].includes(this.has_backdrop)) {
            this.has_backdrop = true;
        }

        $element.once("render", () => {
            const set_controller = variant => {
                if (controller) controller.destroy();

                switch (variant) {
                    case "side"    :
                    case null      :
                    case undefined :
                        controller = new MDSidenavSideController(options);
                        break;
                    case "over" :
                        controller = new MDSidenavOverController(options);
                        break;
                    case "push" :
                        break;
                    default:
                        throw new Error("Invalid variant");
                }
            };

            observer.on("variant", set_controller);
            set_controller(options.variant.toLowerCase());
            options.is_rendered = true;
        });

        observer.on("is_open", is_open => {
            if (controller) controller[is_open ? "open" : "close"]();
        });

        observer.on("position", value => {
            if (controller) controller.change_position(value);
        });
    };

    MDSidenav.prototype.open = function () {
        if (! this.is_open) this.is_open = true;
    };

    MDSidenav.prototype.close = function () {
        if (this.is_open) this.is_open = false;
    };

    MDSidenav.prototype.toggle = function () {
        this.is_open = !this.is_open;
    };
MDSidenav.__jeefo_class__ = true; return MDSidenav;}());

exports.controller_name = "$md_sidenav";
 });
