// @jeefo/material/components/list/index.js
 jeefo.register("node_modules/@jeefo/material/components/list/index.js", async (exports, module) => { const __dirname = "node_modules/@jeefo/material/components/list", __filename = "node_modules/@jeefo/material/components/list/index.js"; const require = path => { return jeefo.require(path, __filename, __dirname); }; /* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
* File Name   : index.js
* Created at  : 2019-10-07
* Updated at  : 2021-03-07
* Author      : jeefo
* Purpose     :
* Description :
* Reference   :
.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.*/
// ignore:start
"use strict";

/* globals*/
/* exported*/

// ignore:end

const md_theme     = (await require("../../services/theme"));
const array_remove = (await require("@jeefo/utils/array/remove"));

const focused_class  = "md-list__item--focused";
const selected_class = "md-list__item--selected";

exports.selector = "md-list";

exports.style = (`.md-list{display:block;padding:8px 0;outline:none}.md-list__item{position:relative;height:48px;cursor:pointer;padding:0 16px;overflow:hidden;display:flex;align-items:center}.md-list__item>.md-icon{margin-right:32px}.md-list__item__meta{margin-left:auto}.md-list__item:focus{outline:none}.md-list__item--selected>.md-ripple:before{opacity:.12}.md-list__link{color:currentColor;text-decoration:none}.md-list--has-leading-icon{padding-left:24px}.md-list--has-leading-icon .md-list__item{padding-left:68px}`
/* space filler






































*/);

/*
md_theme.register_template(`
/* sass /
@import '@jeefo/material'

.md-list__item
    $root: #{&}

    +theme-modifiers($root, (primary, secondary))
        +property-template(color)
`);
*/

exports.template = (`<jf-content select="md-list-item"></jf-content><jf-content select="md-list-link"></jf-content><jf-content select="md-list-divider"></jf-content>`
/* space filler



*/);

exports.controller = (function () { function MDList () {} 
    MDList.prototype.on_init = function ($element) {
        $element.add_class("md-list");
        this.items = [];

        $element.on("initialized", () => {
            if (this.items.length) {
                const found = this.items.find(item => item.is_selected);
                if (found) {
                    this.select(found);
                } else {
                    this.focusable(this.items[0]);
                }
            }
        });
    };

    MDList.prototype.add = function (item) {
        this.items.push(item);
    };

    MDList.prototype.remove = function (item) {
        array_remove(this.items, item);
    };

    MDList.prototype.focusable = function (item) {
        item.is_focusable = true;
        item.$element.set_attr("tabindex", 0);
    };

    MDList.prototype.focus = function (item) {
        this.focusable(item);
        item.$element.add_class(focused_class);
        item.$element.DOM_element.focus();
        for (const _item of this.items) {
            if (_item.is_focusable && _item !== item) {
                _item.is_focusable = false;
                _item.$element.set_attr("tabindex", -1);
            }
        }
    };

    MDList.prototype.select = function (item) {
        this.focus(item);
        item.$element.add_class(selected_class);
        item.is_selected = true;
        for (const _item of this.items) {
            if (_item.is_selected && _item !== item) {
                _item.is_selected = false;
                _item.$element.remove_class(selected_class);
            }
        }
    };
MDList.__jeefo_class__ = true; return MDList;}());

exports.controller_name = "$md_list";
 });
