// @jeefo/material/components/list/item.js
 jeefo.register("node_modules/@jeefo/material/components/list/item.js", async (exports, module) => { const __dirname = "node_modules/@jeefo/material/components/list", __filename = "node_modules/@jeefo/material/components/list/item.js"; const require = path => { return jeefo.require(path, __filename, __dirname); }; /* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
* File Name   : item.js
* Created at  : 2019-10-08
* Updated at  : 2021-03-04
* Author      : jeefo
* Purpose     :
* Description :
* Reference   :
.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.*/
// ignore:start
"use strict";

/* globals*/
/* exported*/

// ignore:end

const IMDListItem = (await require("./event_handler"));

exports.selector = "md-list-item";

exports.template = (`<jf-content select="md-avatar" class="md-list__item__avatar"></jf-content><jf-content select="md-icon" class="md-list__item__support_icon"></jf-content><jf-content class="md-list__item__content"></jf-content><jf-content select="md-list-meta" class="md-list__item__meta"></jf-content><md-ripple></md-ripple>`
/* space filler





*/);

exports.dependencies = {
    md_list_ctrl: "mdList",
};

exports.bindings = {
    is_selected: "<isSelected",
};

exports.controller = (function () { function MDListItem () {}  
    MDListItem.prototype.on_init = function ($element) {
        this.$element = $element;
        IMDListItem.prototype.init.call(this);

        $element.on("click", () => {
            this.md_list_ctrl.select(this);
        });
    };

    MDListItem.prototype.on_destroy = function () {
        this.md_list_ctrl.remove(this);
    };
MDListItem.__jeefo_class__ = true; return MDListItem;}());

exports.controller_name = "$md_list__item";
 });
