// @jeefo/material/components/list/link.js
 jeefo.register("node_modules/@jeefo/material/components/list/link.js", async (exports, module) => { const __dirname = "node_modules/@jeefo/material/components/list", __filename = "node_modules/@jeefo/material/components/list/link.js"; const require = path => { return jeefo.require(path, __filename, __dirname); }; /* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
* File Name   : link.js
* Created at  : 2021-01-19
* Updated at  : 2021-03-07
* Author      : jeefo
* Purpose     :
* Description :
* Reference   :
.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.*/
// ignore:start
"use strict";

/* globals*/
/* exported*/

// ignore:end

const IMDListItem      = (await require("./event_handler"));
const {parse, replace} = (await require("@jeefo/jqlite/dom_parser"));

const anchor = parse(`<a></a>`)[0];
const ripple = parse(`<md-ripple></md-ripple>`)[0];

exports.selector = "md-list-link";

exports.template = element => {
    element.classList.add("md-list__item", "md-list__link");
    element.appendChild(ripple.cloneNode());
    return replace(element, anchor.cloneNode());
};

exports.dependencies = {
    md_list_ctrl: "mdList",
};

exports.bindings = {
    isFocused : '=',
    isSelected: '=',
};

exports.controller = (function () { function MDListLink () {}  
    MDListLink.prototype.on_init = function ($element) {
        this.$element = $element;
        IMDListItem.prototype.init.call(this);

        //$element.on("click", () => { this.md_list_ctrl.select(this); });
    };
MDListLink.__jeefo_class__ = true; return MDListLink;}());
 });
