// @jeefo/material/components/surface/index.js
 jeefo.register("node_modules/@jeefo/material/components/surface/index.js", async (exports, module) => { const __dirname = "node_modules/@jeefo/material/components/surface", __filename = "node_modules/@jeefo/material/components/surface/index.js"; const require = path => { return jeefo.require(path, __filename, __dirname); }; /* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
* File Name   : index.js
* Created at  : 2020-08-01
* Updated at  : 2021-02-24
* Author      : jeefo
* Purpose     :
* Description :
* Reference   :
.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.*/
// ignore:start
"use strict";

/* globals*/
/* exported*/

// ignore:end

const Observer       = (await require("@jeefo/observer"));
const md_theme       = (await require("../../services/theme"));
const class_modifier = (await require("../../utils/class_modifier"));

exports.selector = "md-surface";

exports.style = (`.md-surface{width:100%;height:100%;display:block}`
/* space filler





*/);

md_theme.set_default({
    ".md-surface": {
        "color"            : "$on_surface-color",
        "background-color" : "$surface-color",
    },
    ".md-surface.md-surface--primary": {
        "color"            : "$on_primary-color",
        "background-color" : "$primary-color",
    },
    ".md-surface.md-surface--secondary": {
        "color"            : "$on_secondary-color",
        "background-color" : "$secondary-color",
    },
});

md_theme.register_template((`.md-surface{content:"props:[color, background-color]"}.md-surface.md-surface--primary{content:"props:[color, background-color]"}.md-surface.md-surface--secondary{content:"props:[color, background-color]"}`
/* space filler









*/));

exports.bindings = {
    color: '@',
};

exports.controller = (function () { function MDSurface () {} 
    MDSurface.prototype.on_init = function ($element) {
        const name          = $element.name.toLowerCase();
        const observer      = new Observer(this);
        const modifiers     = ["primary", "secondary", "transparent"];
        const {DOM_element} = $element;

        this.$element = $element;
        $element.add_class("md-surface");

        const on_color_change = class_modifier(DOM_element, name, modifiers);
        observer.on("color", on_color_change);
        on_color_change(this.color);
    };

    MDSurface.prototype.is_closer = function ({DOM_element}) {
        let el = this.$element.DOM_element;
        while ((el = el.parentNode)) {
            if (el === DOM_element) return true;
        }
        return false;
    };
MDSurface.__jeefo_class__ = true; return MDSurface;}());

exports.controller_name = "$md_surface";
 });
