// @jeefo/component/components/invisible_component.js
 jeefo.register("node_modules/@jeefo/component/components/invisible_component.js", async (exports, module) => { const __dirname = "node_modules/@jeefo/component/components", __filename = "node_modules/@jeefo/component/components/invisible_component.js"; const require = path => { return jeefo.require(path, __filename, __dirname); }; /* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
* File Name   : invisible_component.js
* Created at  : 2020-06-08
* Updated at  : 2020-10-23
* Author      : jeefo
* Purpose     :
* Description :
* Reference   :
.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.*/
// ignore:start
"use strict";

/* globals*/
/* exported*/

// ignore:end

const IComponent   = (await require("../interfaces/i_component"));
const array_remove = (await require("@jeefo/utils/array/remove"));

var InvisibleComponent = (function () {  
    function InvisibleComponent (name, definition, parent = null) {
        IComponent.call(this, name, definition);
        this.parent = parent;
        if (parent) parent.children.push(this);
    } InvisibleComponent.prototype = Object.create(IComponent.prototype);

    InvisibleComponent.prototype.initialize = async function () {
        if (this.is_initialized) {
            console.log("InvisibleComponent initialize called more than once");
        }
        for (const child of this.children) {
            if (! child.is_initialized && ! child.is_destroyed) {
                await child.initialize();
            }
            if (this.is_destroyed) return;
        }
        this.is_initialized = true;
    };

    InvisibleComponent.prototype.destroy = async function () {
        if (this.is_destroyed) {
            console.log("InvisibleComponent destroy called more than once.");
        }
        for (const child of this.children) await child.destroy();
        array_remove(this.parent.children, this);
        this.is_destroyed = true;
    };

    InvisibleComponent.prototype.digest = async function () {
        if (this.is_initialized) {
            for (const child of this.children) await child.digest();
        }
    };
InvisibleComponent.__jeefo_class__ = true; return InvisibleComponent;}());

module.exports = InvisibleComponent;
 });
