// @jeefo/component/components/structure_component.js
 jeefo.register("node_modules/@jeefo/component/components/structure_component.js", async (exports, module) => { const __dirname = "node_modules/@jeefo/component/components", __filename = "node_modules/@jeefo/component/components/structure_component.js"; const require = path => { return jeefo.require(path, __filename, __dirname); }; /* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
* File Name   : structure_component.js
* Created at  : 2019-06-26
* Updated at  : 2020-11-24
* Author      : jeefo
* Purpose     :
* Description :
* Reference   :
.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.*/
// ignore:start
"use strict";

/* globals*/
/* exported*/

// ignore:end

const jqlite       = (await require("@jeefo/jqlite"));
const Observer     = (await require("@jeefo/observer"));
const IComponent   = (await require("../interfaces/i_component"));
const array_remove = (await require("@jeefo/utils/array/remove"));

var StructureComponent = (function () {  
    function StructureComponent (name, element, definition, parent) {
        IComponent.call(this, name, definition);

        if (! this.Controller) {
            throw new Error("Controller is required in IStructure");
        }

        this.parent  = parent;
        this.element = element;

        parent.children.push(this);
    } StructureComponent.prototype = Object.create(IComponent.prototype);

    StructureComponent.prototype.initialize = async function () {
        // DEBUG_START
        if (this.is_initialized) {
            console.log("StructureComponent initialize called more than once");
        }
        // DEBUG_END
        this.controller = new this.Controller();
        await this.controller.on_init(jqlite(this.element), this);
        this.is_initialized = true;
    };

    StructureComponent.prototype.destroy = async function () {
        // DEBUG_START
        if (this.is_destroyed) {
            console.log("StructureComponent destroy called more than once.");
        }
        // DEBUG_END
        const {controller, parent, children} = this;
        array_remove(parent.children, this);
        this.is_destroyed = true;

        for (const directive of this.directives) directive.destroy();
        if (controller) {
            if (typeof controller.on_destroy === "function") {
                await controller.on_destroy();
            }
            Observer.destroy(controller);
        }

        let i = children.length;
        while (i--) children[i].destroy();
    };

    StructureComponent.prototype.digest = async function () {
        if (this.is_initialized) {
            if (typeof this.controller.on_digest === "function") {
                await this.controller.on_digest();
            }
            for (const child of this.children) await child.digest();
        }
    };

    StructureComponent.prototype.trigger_render = function () {
        // DEBUG_START
        if (this.is_rendered) {
            console.log("IStructure render called more than once.");
        }
        // DEBUG_END
        for (const child of this.children) child.trigger_render();
        this.is_rendered = true;
    };
StructureComponent.__jeefo_class__ = true; return StructureComponent;}());

module.exports = StructureComponent;
 });
