// @jeefo/component/interfaces/i_component.js
 jeefo.register("node_modules/@jeefo/component/interfaces/i_component.js", async (exports, module) => { const __dirname = "node_modules/@jeefo/component/interfaces", __filename = "node_modules/@jeefo/component/interfaces/i_component.js"; const require = path => { return jeefo.require(path, __filename, __dirname); }; /* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
* File Name   : i_component.js
* Created at  : 2019-07-06
* Updated at  : 2020-11-23
* Author      : jeefo
* Purpose     :
* Description :
* Reference   :
.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.*/
// ignore:start
"use strict";

/* globals*/
/* exported*/

// ignore:end

const Readonly       = (await require("@jeefo/utils/object/readonly"));
const Interface      = (await require("@jeefo/utils/class/interface"));
const virtual_method = (await require("@jeefo/utils/class/virtual_method"));

let static_id = 1;

var IComponent = (function () {  
    function IComponent (name, {
        Controller      = null,
        controller_name = null,
    }, ExtentedInterface, is_directive = false) {
        Interface.call(this, ExtentedInterface || IComponent);

        const readonly = new Readonly(this);
        readonly.getter("name", () => name);

        this.Controller      = Controller;
        this.controller_name = controller_name;
        this.is_rendered     = false;
        this.is_destroyed    = false;
        this.is_initialized  = false;

        if (! is_directive) {
            this.children   = [];
            this.directives = [];
        }

        const id = static_id++;
        readonly.getter("id", () => id);
    } IComponent.prototype = Object.create(Interface.prototype);

    IComponent.prototype.digest = function         () { virtual_method(); };
    IComponent.prototype.destroy = function        () { virtual_method(); };
    IComponent.prototype.initialize = function     () { virtual_method(); };
    IComponent.prototype.trigger_render = function () { virtual_method(); };
IComponent.__jeefo_class__ = true; return IComponent;}());

module.exports = IComponent;
 });
