// @jeefo/material/components/ripple/index.js
 jeefo.register("node_modules/@jeefo/material/components/ripple/index.js", async (exports, module) => { const __dirname = "node_modules/@jeefo/material/components/ripple", __filename = "node_modules/@jeefo/material/components/ripple/index.js"; const require = path => { return jeefo.require(path, __filename, __dirname); }; /* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
* File Name   : index.js
* Created at  : 2020-06-23
* Updated at  : 2021-02-26
* Author      : jeefo
* Purpose     :
* Description :
* Reference   :
.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.*/
// ignore:start
"use strict";
/* globals*/
/* exported*/

const WAVE_CLASS           = "md-ripple__wave";
const ACTIVATED_WAVE_CLASS = "md-ripple__wave--activated";
const REMOVABLE_WAVE_CLASS = "md-ripple__wave--removable";

// ignore:end

const {max, sqrt} = Math;

module.exports = {
    selector : "md-ripple",
    style : (`.md-ripple{width:100%;height:100%;position:absolute;top:0;left:0;overflow:hidden;will-change:transform}.md-ripple:before{width:100%;height:100%;position:absolute;top:0;left:0;content:"";opacity:0;z-index:0;transition:opacity .55s cubic-bezier(0.25, 0.8, 0.25, 1);pointer-events:none;background-color:currentColor}:not(:disabled):hover>.md-ripple:before{opacity:.05}:not(:disabled):focus>.md-ripple:before{opacity:.08}.md-ripple__wave{opacity:.11;position:absolute;transform:translate(-50%, -50%) scale(0);border-radius:50%;transform-origin:50% 50%;background-color:currentColor;transition-duration:.45s;transition-property:opacity,transform;transition-timing-function:cubic-bezier(0.25, 0.8, 0.25, 1)}.md-ripple__wave--activated{transform:translate(-50%, -50%) scale(1)}`
/* space filler





*/),
    template (element) {
        if (element.firstChild) {
            throw new Error("MDRipple shouldn't contain any child");
        }
        element.classList.add("md-ripple");
    },
    bindings : {
    },
    controller : (function () { function MDRipple () {} 
        MDRipple.prototype.on_init = function ($element) {
            let active_ripple;

            const on_mouse_up = function () {
                if (event.button === 0 && active_ripple) {
                    if (active_ripple.is_removable) {
                        $element.DOM_element.removeChild(active_ripple);
                    } else {
                        active_ripple.is_removable = true;
                    }
                    active_ripple = null;
                }
            };

            $element.on("mouseup"    , on_mouse_up);
            $element.on("mouseleave" , on_mouse_up);

            const on_mousedown = event => {
                if (event.button !== 0) { return; }

                let x, y;
                if (event.srcElement === $element.DOM_element) {
                    x = event.offsetX;
                    y = event.offsetY;
                } else {
                    const {left, top} = $element.rect();
                    x = event.clientX - left;
                    y = event.clientY - top;
                }

                const width    = max($element.width  - x, x);
                const height   = max($element.height - y, y);
                const diameter = sqrt(width*width + height*height) * 2;

                const ripple = document.createElement("div");
                ripple.style.top    = `${y}px`;
                ripple.style.left   = `${x}px`;
                ripple.style.width  = `${diameter}px`;
                ripple.style.height = `${diameter}px`;
                ripple.classList.add(WAVE_CLASS);

                $element.append(ripple);
                ripple.getBoundingClientRect();
                ripple.classList.add(ACTIVATED_WAVE_CLASS);
                active_ripple = ripple;

                const {transitionDuration} = getComputedStyle(active_ripple);
                setTimeout(() => {
                    if (! $element.DOM_element) return;
                    if (ripple.is_removable) {
                        $element.DOM_element.removeChild(ripple);
                    } else {
                        ripple.is_removable = true;
                    }
                }, parseFloat(transitionDuration) * 1000);
            };

            $element.on("mousedown" , on_mousedown);
            $element.on("touchstart", on_mousedown);
        };
    MDRipple.__jeefo_class__ = true; return MDRipple;}())
};
 });
