// @jeefo/observer/index.js
 jeefo.register("node_modules/@jeefo/observer/index.js", async (exports, module) => { const __dirname = "node_modules/@jeefo/observer", __filename = "node_modules/@jeefo/observer/index.js"; const require = path => { return jeefo.require(path, __filename, __dirname); }; /* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
* File Name   : index.js
* Created at  : 2017-08-05
* Updated at  : 2020-10-22
* Author      : jeefo
* Purpose     :
* Description :
.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.*/
// ignore:start
"use strict";

/* globals*/
/* exported*/

/*
 * TODO: make observerable deep nested property
 * example:
 * 	`observer.on("options.analyser.disabled", is_disabled => {...})`
 */

// ignore:end

const array_remove     = (await require("@jeefo/utils/array/remove"));
const {defineProperty} = Object;

const observer_instances = [];

var JeefoStaticObserver = (function () { 
    function JeefoStaticObserver (object) {
        this.object    = object;
        this.values    = {};
        this.notifiers = [];
    }
JeefoStaticObserver.__jeefo_class__ = true; return JeefoStaticObserver;}());

var JeefoObserver = (function () { 
    function JeefoObserver (object) {
        this.object = object;

        const private_instance = observer_instances.find(instance => {
            return instance.object === object;
        });
        if (! private_instance) {
            observer_instances.push(new JeefoStaticObserver(object));
        }
    }

    JeefoObserver.prototype.on = function (property, notify_handler) {
        const private_instance = observer_instances.find(instance => {
            return instance.object === this.object;
        });
        if (! private_instance) { return; }

        const { object, values, notifiers } = private_instance;

        if (notifiers[property]) {
            notifiers[property].push(notify_handler);
        } else {
            values[property]    = object[property];
            notifiers[property] = [notify_handler];

            defineProperty(object, property, {
                get () { return values[property]; },

                set (value) {
                    const old_value = values[property];
                    if (old_value !== value) {
                        values[property] = value;
                        notifiers[property].forEach(fn => fn(value, old_value));
                    }
                }
            });
        }

        return notify_handler;
    };

    JeefoObserver.prototype.off = function (property, notify_handler) {
        const private_instance = observer_instances.find(instance => {
            return instance.object === this.object;
        });
        if (! private_instance) { return; }

        const { object, notifiers } = private_instance;

        array_remove(notifiers[property], notify_handler);
        if (notifiers[property].length === 0) {
            const value = object[property];
            delete object[property];
            object[property] = value;
        }
    };

    JeefoObserver.prototype.once = function (property, notify_handler) {
        const notifier = this.on(property, new_value => {
            notify_handler(new_value);
            this.off(property, notifier);
        });
    };

    JeefoObserver.destroy = function (object) {
        const private_instance = observer_instances.find(instance => {
            return instance.object === object;
        });
        if (private_instance) {
            array_remove(observer_instances, private_instance);
        }
    };
JeefoObserver.__jeefo_class__ = true; return JeefoObserver;}());

module.exports = JeefoObserver;
 });
