// @jeefo/tokenizer/cursor_position.js
 jeefo.register("node_modules/@jeefo/tokenizer/cursor_position.js", async (exports, module) => { const __dirname = "node_modules/@jeefo/tokenizer", __filename = "node_modules/@jeefo/tokenizer/cursor_position.js"; const require = path => { return jeefo.require(path, __filename, __dirname); }; /* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
* File Name   : cursor_position.js
* Created at  : 2019-03-15
* Updated at  : 2019-06-23
* Author      : jeefo
* Purpose     :
* Description :
* Reference   :
.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.*/
// ignore:start
"use strict";

/* globals*/
/* exported*/

// ignore:end

var CursorPosition = (function () { 
    function CursorPosition (cursor_position) {
        this.index          = cursor_position.index;
        this.line           = cursor_position.line;
        this.column         = cursor_position.column;
        this.virtual_column = cursor_position.virtual_column;
    }

    CursorPosition.prototype.clone = function () {
		return new CursorPosition(this);
    };
CursorPosition.__jeefo_class__ = true; return CursorPosition;}());

module.exports = CursorPosition;
 });
