// @jeefo/tokenizer/cursor.js
 jeefo.register("node_modules/@jeefo/tokenizer/cursor.js", async (exports, module) => { const __dirname = "node_modules/@jeefo/tokenizer", __filename = "node_modules/@jeefo/tokenizer/cursor.js"; const require = path => { return jeefo.require(path, __filename, __dirname); }; /* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
* File Name   : cursor.js
* Created at  : 2019-03-15
* Updated at  : 2019-06-28
* Author      : jeefo
* Purpose     :
* Description :
* Reference   :
.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.*/
// ignore:start
"use strict";

/* globals*/
/* exported*/

const MAX_HISTORY_LENGTH = 10;

// ignore:end

const CursorPosition = (await require("./cursor_position"));

var Cursor = (function () { 
    function Cursor () {
        this.position = new CursorPosition({
            index          : -1,
            line           : 1,
            column         : 0,
            virtual_column : 0
        });

        this.history = [];
    }

    Cursor.prototype.has_saved_position = function () {
        return this.history.length > 0;
    };

    Cursor.prototype.move_next = function () {
		this.position.index          += 1;
		this.position.column         += 1;
		this.position.virtual_column += 1;
    };

    Cursor.prototype.move = function (length, virtual_length) {
		this.position.index          += length;
		this.position.column         += length;
		this.position.virtual_column += virtual_length || length;
    };

    Cursor.prototype.save = function () {
        this.history.push(this.position.clone());
        if (this.history.length === MAX_HISTORY_LENGTH) {
            throw new Error("Exceeded max cursor positions history");
        }
    };

    Cursor.prototype.commit = function () {
        if (this.history.length) {
            this.history.pop();
        } else {
            throw new Error("Commiting empty cursor position history");
        }
    };

    Cursor.prototype.rollback = function () {
        if (this.history.length) {
            this.position = this.history.pop();
        } else {
            throw new Error("Rollback empty cursor position history");
        }
    };
Cursor.__jeefo_class__ = true; return Cursor;}());

module.exports = Cursor;
 });
