// @jeefo/component/components/conditional_component.js
 jeefo.register("node_modules/@jeefo/component/components/conditional_component.js", async (exports, module) => { const __dirname = "node_modules/@jeefo/component/components", __filename = "node_modules/@jeefo/component/components/conditional_component.js"; const require = path => { return jeefo.require(path, __filename, __dirname); }; /* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
* File Name   : conditional_component.js
* Created at  : 2020-06-09
* Updated at  : 2020-10-31
* Author      : jeefo
* Purpose     :
* Description :
* Reference   :
.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.*/
// ignore:start
"use strict";

/* globals*/
/* exported*/

// ignore:end

const jqlite         = (await require("@jeefo/jqlite"));
const JeefoDOMParser = (await require("@jeefo/jqlite/dom_parser"));
const array_remove   = (await require("@jeefo/utils/array/remove"));
const compile        = (await require("../compiler"));
const IComponent     = (await require("../interfaces/i_component"));

var ConditionalComponent = (function () {  
    function ConditionalComponent (name, element, parent = null, definition = {}) {
        IComponent.call(this, name, definition);
        this.parent = parent;

        return new Promise(async resolve => {
            const dom_parser = new JeefoDOMParser('');
            const wrapper = dom_parser.document.body;

            const clone = element.cloneNode(true);
            wrapper.appendChild(clone);
            await compile.from_elements([clone], this, false);
            this.$element = jqlite(wrapper.removeChild(wrapper.firstChild));

            if (parent) parent.children.push(this);
            resolve(this);
        });
    } ConditionalComponent.prototype = Object.create(IComponent.prototype);

    ConditionalComponent.prototype.initialize = async function () {
        // DEBUG_START
        if (this.is_initialized) {
            console.log("ConditionalComponent initialize called more than once");
        }
        // DEBUG_END
        for (const child of this.children) {
            await child.initialize();
        }
        this.is_initialized = true;
    };

    ConditionalComponent.prototype.destroy = async function () {
        // DEBUG_START
        if (this.is_destroyed) {
            console.log("ConditionalComponent destroy called more than once.");
        }
        // DEBUG_END
        for (const child of this.children) {
            await child.destroy();
        }
        if (this.$element) {
            this.$element.remove();
        }
        if (this.parent) {
            array_remove(this.parent.children, this);
        }
        this.is_destroyed = true;
    };

    ConditionalComponent.prototype.digest = async function () {
        if (this.is_initialized) {
            for (const child of this.children) {
                await child.digest();
            }
        }
    };

    ConditionalComponent.prototype.trigger_initialized = function () {
        // DEBUG_START
        if (! this.is_initialized) {
            console.log("ConditionalComponent render called more than once.");
        }
        // DEBUG_END
        for (const child of this.children) {
            child.trigger_initialized();
        }
        this.is_rendered = true;
    };

    ConditionalComponent.prototype.trigger_render = function () {
        // DEBUG_START
        if (! this.is_initialized) {
            console.log("ConditionalComponent render called more than once.");
        }
        // DEBUG_END
        for (const child of this.children) {
            child.trigger_render();
        }
        this.is_rendered = true;
    };
ConditionalComponent.__jeefo_class__ = true; return ConditionalComponent;}());

module.exports = ConditionalComponent;
 });
