// @jeefo/monkey_patcher/index.js
 jeefo.register("node_modules/@jeefo/monkey_patcher/index.js", async (exports, module) => { const __dirname = "node_modules/@jeefo/monkey_patcher", __filename = "node_modules/@jeefo/monkey_patcher/index.js"; const require = path => { return jeefo.require(path, __filename, __dirname); }; /* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
* File Name   : index.js
* Created at  : 2019-07-03
* Updated at  : 2019-09-13
* Author      : jeefo
* Purpose     :
* Description :
* Reference   :
.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.*/
// ignore:start
"use strict";

/* globals*/
/* exported*/

// ignore:end

const to_string = function () {
    return Function.prototype.toString.toString();
};
to_string.toString = to_string;

var MonkeyPatcher = (function () { 
    function MonkeyPatcher (object, method_name, patcher) {
        const self            = this;
        const original_method = object[method_name];

        const _patcher = function () {
            return patcher.call(self, original_method, this, arguments);
        };

        const toString = function () {
            return original_method.toString();
        };

        toString.toString = to_string;
        _patcher.toString = toString;

        object[method_name] = _patcher;
    }

    MonkeyPatcher.prototype.patch = function (fn) {
        const self = this;
        return function () {
			return self.run(fn, this, arguments);
        };
    };

	MonkeyPatcher.prototype.run = function (fn, context, args) {
        let result;
		try {
			this.on_enter();
			result = fn.apply(context, args);
		} catch (e) {
			this.on_error(e);
		} finally {
			this.on_leave();
            return result;
		}
	};

    MonkeyPatcher.prototype.on_enter = function () {};
    MonkeyPatcher.prototype.on_error = function () {};
    MonkeyPatcher.prototype.on_leave = function () {};
MonkeyPatcher.__jeefo_class__ = true; return MonkeyPatcher;}());

module.exports = MonkeyPatcher;
 });
