// @jeefo/material/components/divider.js
 jeefo.register("node_modules/@jeefo/material/components/divider.js", async (exports, module) => { const __dirname = "node_modules/@jeefo/material/components", __filename = "node_modules/@jeefo/material/components/divider.js"; const require = path => { return jeefo.require(path, __filename, __dirname); }; /* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
* File Name   : divider.js
* Created at  : 2020-10-18
* Updated at  : 2021-03-11
* Author      : jeefo
* Purpose     :
* Description :
* Reference   :
.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.*/
// ignore:start
"use strict";

/* globals*/
/* exported*/

// ignore:end

const theme_service = (await require("../services/theme"));

exports.selector = "md-divider";

exports.style = (`.md-divider{width:100%;height:1px;display:block;flex-shrink:0}.md-divider--vertical{width:1px;height:100%}`
/* space filler









*/);

theme_service.set_default({
    ".md-divider": {
        "background-color": "$divider-color"
    }
});

theme_service.register_template((`.md-divider{content:"props:[background-color]"}`
/* space filler




*/));

exports.bindings = {
    orientation : "@",
};

exports.controller = (function () { function MDDivider () {} 
    MDDivider.prototype.on_init = function ($element) {
        $element.add_class("md-divider");

        if (typeof this.orientation === "string") {
            const orientation = this.orientation.toLowerCase();
            if (orientation === "vertical") {
                $element.add_class("md-divider--vertical");
            }
        }
    };
MDDivider.__jeefo_class__ = true; return MDDivider;}());

exports.controller_name = "$md_divider";
 });
