// @jeefo/jqlite/jeefo_element.js
 jeefo.register("node_modules/@jeefo/jqlite/jeefo_element.js", async (exports, module) => { const __dirname = "node_modules/@jeefo/jqlite", __filename = "node_modules/@jeefo/jqlite/jeefo_element.js"; const require = path => { return jeefo.require(path, __filename, __dirname); }; /* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
* File Name   : jeefo_element.js
* Created at  : 2017-01-06
* Updated at  : 2020-10-21
* Author      : jeefo
* Purpose     :
* Description :
.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.*/
// ignore:start
"use strict";

/* globals*/
/* exported*/

/*
NODE_TYPE_ELEMENT           = 1,
NODE_TYPE_ATTRIBUTE         = 2,
NODE_TYPE_TEXT              = 3,
NODE_TYPE_COMMENT           = 8,
NODE_TYPE_DOCUMENT          = 9,
NODE_TYPE_DOCUMENT_FRAGMENT = 11

['ELEMENT_NODE', 1],
['ATTRIBUTE_NODE', 2],
['TEXT_NODE', 3],
['CDATA_SECTION_NODE', 4],
['ENTITY_REFERENCE_NODE', 5],
['ENTITY_NODE', 6],
['PROCESSING_INSTRUCTION_NODE', 7],
['COMMENT_NODE', 8],
['DOCUMENT_NODE', 9],
['DOCUMENT_TYPE_NODE', 10],
['DOCUMENT_FRAGMENT_NODE', 11],
['NOTATION_NODE', 12]
*/

//ignore:end

const dash_case = (await require("@jeefo/utils/string/dash_case"));
const {slice}   = Array.prototype;

// Constructor
var JeefoElement = (function () { 
    function JeefoElement (element) {
        this.DOM_element = element;
    }

    (function () { var __JEEFO_PROPERTY_НЭР__ = "value"; var __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__ = Object.getOwnPropertyDescriptor(JeefoElement.prototype, __JEEFO_PROPERTY_НЭР__) || { configurable: true, enumerable: false }; __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__.get = function ()      { return this.DOM_element.value;  }; Object.defineProperty(JeefoElement.prototype, __JEEFO_PROPERTY_НЭР__, __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__); }());
    (function () { var __JEEFO_PROPERTY_НЭР__ = "value"; var __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__ = Object.getOwnPropertyDescriptor(JeefoElement.prototype, __JEEFO_PROPERTY_НЭР__) || { configurable: true, enumerable: false }; __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__.set = function (value) { this.DOM_element.value = value; }; Object.defineProperty(JeefoElement.prototype, __JEEFO_PROPERTY_НЭР__, __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__); }());

    // Don't use innerText here.
    // Because innerText takes CSS styles into account, triggering reflow to
    // up-to-date computes style and computationally expensive.
	(function () { var __JEEFO_PROPERTY_НЭР__ = "text"; var __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__ = Object.getOwnPropertyDescriptor(JeefoElement.prototype, __JEEFO_PROPERTY_НЭР__) || { configurable: true, enumerable: false }; __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__.get = function ()      { return this.DOM_element.textContent;  }; Object.defineProperty(JeefoElement.prototype, __JEEFO_PROPERTY_НЭР__, __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__); }());
	(function () { var __JEEFO_PROPERTY_НЭР__ = "text"; var __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__ = Object.getOwnPropertyDescriptor(JeefoElement.prototype, __JEEFO_PROPERTY_НЭР__) || { configurable: true, enumerable: false }; __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__.set = function (value) { this.DOM_element.textContent = value; }; Object.defineProperty(JeefoElement.prototype, __JEEFO_PROPERTY_НЭР__, __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__); }());

    (function () { var __JEEFO_PROPERTY_НЭР__ = "name"; var __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__ = Object.getOwnPropertyDescriptor(JeefoElement.prototype, __JEEFO_PROPERTY_НЭР__) || { configurable: true, enumerable: false }; __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__.get = function () { return this.DOM_element.tagName; }; Object.defineProperty(JeefoElement.prototype, __JEEFO_PROPERTY_НЭР__, __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__); }());

	// DOM methods
    JeefoElement.prototype.is = function (node) {
        return this.DOM_element === node;
    };
	JeefoElement.prototype.remove = function () {
        if (this.DOM_element && this.DOM_element.parentNode) {
            this.DOM_element.parentNode.removeChild(this.DOM_element);
        }
        this.DOM_element = null;
        return this;
	};
	JeefoElement.prototype.before = function (node) {
        if (node instanceof JeefoElement) { node = node.DOM_element; }
		this.DOM_element.parentNode.insertBefore(node, this.DOM_element);
        return this;
	};
	JeefoElement.prototype.after = function (node) {
        if (node instanceof JeefoElement) { node = node.DOM_element; }
		this.DOM_element.parentNode.insertBefore(
            node, this.DOM_element.nextSibling
        );
        return this;
	};
	JeefoElement.prototype.clone = function (is_deep) {
        return new JeefoElement(this.DOM_element.cloneNode(is_deep));
	};
    JeefoElement.prototype.parent = function () {
        if (this.DOM_element.parentNode) {
            return new JeefoElement(this.DOM_element.parentNode);
        }
        return null;
    };
	JeefoElement.prototype.append = function (node) {
        if (node instanceof JeefoElement) { node = node.DOM_element; }
		this.DOM_element.appendChild(node);
        return this;
	};
	JeefoElement.prototype.replace_with = function (node) {
        if (node instanceof JeefoElement) { node = node.DOM_element; }
		this.DOM_element.replaceWith(node);
        this.DOM_element = node;
        return this;
	};
	JeefoElement.prototype.replace = function (node) {
        if (node instanceof JeefoElement) { node = node.DOM_element; }
		this.DOM_element.parentNode.replaceChild(node, this.DOM_element);
        this.DOM_element = node;
        return this;
	};

	// Selector methods
	JeefoElement.prototype.first = function (query) {
        const element = this.DOM_element.querySelector(query);
        return element ? new JeefoElement(element) : null;
	};
	JeefoElement.prototype.find = function (query) {
        const elements = this.DOM_element.querySelectorAll(query);
        return slice.call(elements);
	};
	JeefoElement.prototype.children = function (index, ...args) {
        let result = this.DOM_element.children[index];
        for (const i of args) {
            result = result.children[i];
        }
		return new JeefoElement(result);
	};
    JeefoElement.prototype.prev = function () {
		return new JeefoElement(this.DOM_element.previousElementSibling);
    };
    JeefoElement.prototype.next = function () {
		return new JeefoElement(this.DOM_element.nextElementSibling);
    };

	// Attribute methods
	JeefoElement.prototype.get_attr = function (key, to_dash_case = true) {
		return this.DOM_element.getAttribute(
            to_dash_case ? dash_case(key) : key
        );
	};
	JeefoElement.prototype.set_attr = function (key, value = '', to_dash_case = true) {
		this.DOM_element.setAttribute(
            to_dash_case ? dash_case(key) : key, value
        );
        return this;
	};
	JeefoElement.prototype.has_attr = function (key, to_dash_case = true) {
		return this.DOM_element.hasAttribute(
            to_dash_case ? dash_case(key) : key
        );
	};
	JeefoElement.prototype.remove_attr = function (key, to_dash_case = true) {
		this.DOM_element.removeAttribute(
            to_dash_case ? dash_case(key) : key
        );
        return this;
	};
    JeefoElement.prototype.attrs = function (attrs, to_dash_case = true) {
        for (const prop of Object.keys(attrs)) {
            this.DOM_element.setAttribute(
                to_dash_case ? dash_case(prop) : prop, attrs[prop]
            );
        }
        return this;
    };
JeefoElement.__jeefo_class__ = true; return JeefoElement;}());

(await require("./event_methods"))(JeefoElement);
(await require("./class_methods"))(JeefoElement);
(await require("./style_methods"))(JeefoElement);

module.exports = JeefoElement;
 });
