// @jeefo/tokenizer/index.js
 jeefo.register("node_modules/@jeefo/tokenizer/index.js", async (exports, module) => { const __dirname = "node_modules/@jeefo/tokenizer", __filename = "node_modules/@jeefo/tokenizer/index.js"; const require = path => { return jeefo.require(path, __filename, __dirname); }; /* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
* File Name   : index.js
* Created at  : 2017-04-12
* Updated at  : 2019-10-31
* Author      : jeefo
* Purpose     :
* Description :
.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.*/
// ignore:start
"use strict";

/* globals*/
/* exported*/

//ignore:end

const Interface       = (await require("@jeefo/utils/class/interface"));
const StringStream    = (await require("./string_stream"));
const TokenDefinition = (await require("./token_definition"));

var ITokenDefinition = (function () {  
    function ITokenDefinition () { Interface.call(this, ITokenDefinition); } ITokenDefinition.prototype = Object.create(Interface.prototype);
ITokenDefinition.__jeefo_class__ = true; return ITokenDefinition;}());

const sort_by_priority = (a, b) => {
    return a.Token.prototype.priority - b.Token.prototype.priority;
};

var JeefoTokenizer = (function () { 
    function JeefoTokenizer () {
        this.streamer          = new StringStream('');
        this.token_definitions = [];
    }

    JeefoTokenizer.prototype.init = function (source_code, tab_space) {
        this.streamer = new StringStream(source_code, tab_space);
    };

    JeefoTokenizer.prototype.get_next_token = function () {
        const current_character = this.streamer.next(true);

        if (current_character === null) { return null; }

        let i = this.token_definitions.length;
        while (i--) {
            const token_def = this.token_definitions[i];
            if (token_def.is(current_character, this.streamer)) {
                const token = new token_def.Token();
                token_def.initialize(
                    token, current_character, this.streamer
                );

                return token;
            }
        }

        throw new SyntaxError("Undefined token");
    };

    JeefoTokenizer.prototype.register = function (token_definition) {
        // TODO: investigate what is this...
        // is that good idea or bad idea ???
        if (! (token_definition instanceof ITokenDefinition)) {
            // throw Error
        }
        this.token_definitions.push(new TokenDefinition(token_definition));
        this.token_definitions.sort(sort_by_priority);

        return this;
    };
JeefoTokenizer.__jeefo_class__ = true; return JeefoTokenizer;}());

module.exports = JeefoTokenizer;
 });
