// @jeefo/tokenizer/token_definition.js
 jeefo.register("node_modules/@jeefo/tokenizer/token_definition.js", async (exports, module) => { const __dirname = "node_modules/@jeefo/tokenizer", __filename = "node_modules/@jeefo/tokenizer/token_definition.js"; const require = path => { return jeefo.require(path, __filename, __dirname); }; /* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
* File Name   : token_definition.js
* Created at  : 2017-08-10
* Updated at  : 2019-07-18
* Author      : jeefo
* Purpose     :
* Description :
.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.*/
// ignore:start
"use strict";

/* globals*/
/* exported*/

// ignore:end

var TokenDefinition = (function () { 
    function TokenDefinition (token_definition) {
        if (token_definition === null || typeof token_definition !== "object") {
            throw new Error("Invalid argument: token_definition");
        }
        if (typeof token_definition.id !== "string") {
            throw new Error("Invalid argument: token_definition.id");
        }
        if (typeof token_definition.priority !== "number") {
            throw new Error("Invalid argument: token_definition.priority");
        }
        if (typeof token_definition.is !== "function") {
            throw new Error("Invalid argument: token_definition.is");
        }
        if (typeof token_definition.initialize !== "function") {
            throw new Error("Invalid argument: token_definition.initialize");
        }

        function Token () {}
        Token.prototype.id       = token_definition.id;
        Token.prototype.priority = token_definition.priority;

        this.is         = token_definition.is;
        this.Token      = Token;
        this.initialize = token_definition.initialize;
    }
TokenDefinition.__jeefo_class__ = true; return TokenDefinition;}());

module.exports = TokenDefinition;
 });
