// @jeefo/material/components/selection/index.js
 jeefo.register("node_modules/@jeefo/material/components/selection/index.js", async (exports, module) => { const __dirname = "node_modules/@jeefo/material/components/selection", __filename = "node_modules/@jeefo/material/components/selection/index.js"; const require = path => { return jeefo.require(path, __filename, __dirname); }; /* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
* File Name   : index.js
* Created at  : 2021-02-02
* Updated at  : 2021-02-26
* Author      : jeefo
* Purpose     :
* Description :
* Reference   :
.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.*/
// ignore:start
"use strict";

/* globals*/
/* exported*/

// ignore:end

let id = 0;

exports.style = (`.md-selection{display:flex;padding:8px;user-select:none;align-items:center}.md-selection label{cursor:pointer}.md-selection .md-radio+label,.md-selection .md-checkbox+label{margin-left:12px}.md-selection--baseline{align-items:baseline}`
/* space filler


















*/);

exports.template = (`<jf-content select="md-radio"></jf-content><jf-content select="md-checkbox"></jf-content><jf-content select="label" md-emphasis="{{ $md_selection.emphasis }}"></jf-content>`
/* space filler






*/);

exports.bindings = {
    align: '@',
};

exports.controller = (function () { function MDSelection () {} 
    MDSelection.prototype.on_init = function ($element) {
        const $input  = $element.first("input");
        const input   = $input.DOM_element;
        const $label  = $element.first("label");
        this.emphasis = "medium";
        $element.add_class("md-selection");

        const set_emphasis = () => {
            this.emphasis = input.disabled ? "disabled"
                : input.checked ? "high" : "medium";
        };

        $element.on("initialized", set_emphasis);

        if ($label && $input) {
            id += 1;
            const _id = `md-selection-${id}`;
            $input.set_attr("id"  , _id);
            $label.set_attr("for" , _id);
            $label.on("mousedown", () => $input.trigger("mousedown"));

            $input.on("change", set_emphasis);
            $input.on("state_changed", set_emphasis);
        }

        if (this.align) $element.add_class(`md-selection--${this.align}`);
    };
MDSelection.__jeefo_class__ = true; return MDSelection;}());

exports.controller_name = "$md_selection";
 });
