// jeefo/.caches/app/main.js
 jeefo.register("./main.js", async (exports, module) => { const __dirname = ".", __filename = "./main.js"; const require = path => { return jeefo.require(path, __filename, __dirname); }; /* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
* File Name   : main.js
* Created at  : 2020-12-28
* Updated at  : 2021-01-06
* Author      : jeefo
* Purpose     :
* Description :
* Reference   :
.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.*/
// ignore:start
"use strict";

/* globals*/
/* exported*/

// ignore:end

/*
jeefo.load_script("/js/test.js");
jeefo.promise_all([/* imaginary promises /]);
*/

(await require("./polyfills"));

const compile              = (await require("@jeefo/component/compiler"));
const {InvisibleComponent} = (await require("@jeefo/component"));

try {
    function A () {}
    var B = (function () {  
        function B () {
            A.call(this);
        } B.prototype = Object.create(A.prototype);

        B.prototype[123] = function () {};
        B.prototype['str'] = function () {};
        B.prototype.someMethod = function () {};
        B.prototype['computed'] = function () {};

        (function () { var __JEEFO_PROPERTY_НЭР__ = "m"; var __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__ = Object.getOwnPropertyDescriptor(B.prototype, __JEEFO_PROPERTY_НЭР__) || { configurable: true, enumerable: false }; __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__.get = function () {}; Object.defineProperty(B.prototype, __JEEFO_PROPERTY_НЭР__, __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__); }());
        (function () { var __JEEFO_PROPERTY_НЭР__ = 0; var __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__ = Object.getOwnPropertyDescriptor(B.prototype, __JEEFO_PROPERTY_НЭР__) || { configurable: true, enumerable: false }; __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__.get = function () {}; Object.defineProperty(B.prototype, __JEEFO_PROPERTY_НЭР__, __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__); }());
        (function () { var __JEEFO_PROPERTY_НЭР__ = 'str_'; var __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__ = Object.getOwnPropertyDescriptor(B.prototype, __JEEFO_PROPERTY_НЭР__) || { configurable: true, enumerable: false }; __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__.get = function () {}; Object.defineProperty(B.prototype, __JEEFO_PROPERTY_НЭР__, __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__); }());
        (function () { var __JEEFO_PROPERTY_НЭР__ = 1+2; var __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__ = Object.getOwnPropertyDescriptor(B.prototype, __JEEFO_PROPERTY_НЭР__) || { configurable: true, enumerable: false }; __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__.get = function () {}; Object.defineProperty(B.prototype, __JEEFO_PROPERTY_НЭР__, __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__); }());

        (function () { var __JEEFO_PROPERTY_НЭР__ = "m"; var __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__ = Object.getOwnPropertyDescriptor(B.prototype, __JEEFO_PROPERTY_НЭР__) || { configurable: true, enumerable: false }; __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__.set = function (v) {return v;}; Object.defineProperty(B.prototype, __JEEFO_PROPERTY_НЭР__, __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__); }());
        (function () { var __JEEFO_PROPERTY_НЭР__ = 0; var __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__ = Object.getOwnPropertyDescriptor(B.prototype, __JEEFO_PROPERTY_НЭР__) || { configurable: true, enumerable: false }; __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__.set = function (v) {return v;}; Object.defineProperty(B.prototype, __JEEFO_PROPERTY_НЭР__, __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__); }());
        (function () { var __JEEFO_PROPERTY_НЭР__ = 'str_'; var __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__ = Object.getOwnPropertyDescriptor(B.prototype, __JEEFO_PROPERTY_НЭР__) || { configurable: true, enumerable: false }; __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__.set = function (v) {return v;}; Object.defineProperty(B.prototype, __JEEFO_PROPERTY_НЭР__, __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__); }());
        (function () { var __JEEFO_PROPERTY_НЭР__ = 1+2; var __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__ = Object.getOwnPropertyDescriptor(B.prototype, __JEEFO_PROPERTY_НЭР__) || { configurable: true, enumerable: false }; __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__.set = function (v) {return v;}; Object.defineProperty(B.prototype, __JEEFO_PROPERTY_НЭР__, __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__); }());

        B.prototype[1234] = async function () {};
        B.prototype['async_str'] = async function () {};
        B.prototype.asyncAsyncMethod = async function () {};
        B.prototype['async_computed'] = async function () {};

        B[123] = function () {};
        B['static_str'] = function () {};
        B.someStaticMethod = function () {};
        B['static_computed'] = function () {};

        B[1234] = async function () {};
        B['async_str'] = async function () {};
        B.someStaticAsyncMethod = async function () {};
        B['async_computed'] = async function () {};

        (function () { var __JEEFO_PROPERTY_НЭР__ = "m"; var __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__ = Object.getOwnPropertyDescriptor(B, __JEEFO_PROPERTY_НЭР__) || { configurable: true, enumerable: false }; __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__.get = function () {}; Object.defineProperty(B, __JEEFO_PROPERTY_НЭР__, __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__); }());
        (function () { var __JEEFO_PROPERTY_НЭР__ = 3.14; var __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__ = Object.getOwnPropertyDescriptor(B, __JEEFO_PROPERTY_НЭР__) || { configurable: true, enumerable: false }; __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__.get = function () {}; Object.defineProperty(B, __JEEFO_PROPERTY_НЭР__, __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__); }());
        (function () { var __JEEFO_PROPERTY_НЭР__ = 'str'; var __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__ = Object.getOwnPropertyDescriptor(B, __JEEFO_PROPERTY_НЭР__) || { configurable: true, enumerable: false }; __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__.get = function () {}; Object.defineProperty(B, __JEEFO_PROPERTY_НЭР__, __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__); }());
        (function () { var __JEEFO_PROPERTY_НЭР__ = 1+2; var __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__ = Object.getOwnPropertyDescriptor(B, __JEEFO_PROPERTY_НЭР__) || { configurable: true, enumerable: false }; __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__.get = function () {}; Object.defineProperty(B, __JEEFO_PROPERTY_НЭР__, __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__); }());

        (function () { var __JEEFO_PROPERTY_НЭР__ = "m"; var __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__ = Object.getOwnPropertyDescriptor(B, __JEEFO_PROPERTY_НЭР__) || { configurable: true, enumerable: false }; __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__.set = function (v) {return v;}; Object.defineProperty(B, __JEEFO_PROPERTY_НЭР__, __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__); }());
        (function () { var __JEEFO_PROPERTY_НЭР__ = 3.14; var __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__ = Object.getOwnPropertyDescriptor(B, __JEEFO_PROPERTY_НЭР__) || { configurable: true, enumerable: false }; __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__.set = function (v) {return v;}; Object.defineProperty(B, __JEEFO_PROPERTY_НЭР__, __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__); }());
        (function () { var __JEEFO_PROPERTY_НЭР__ = 'str'; var __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__ = Object.getOwnPropertyDescriptor(B, __JEEFO_PROPERTY_НЭР__) || { configurable: true, enumerable: false }; __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__.set = function (v) {return v;}; Object.defineProperty(B, __JEEFO_PROPERTY_НЭР__, __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__); }());
        (function () { var __JEEFO_PROPERTY_НЭР__ = 1+2; var __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__ = Object.getOwnPropertyDescriptor(B, __JEEFO_PROPERTY_НЭР__) || { configurable: true, enumerable: false }; __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__.set = function (v) {return v;}; Object.defineProperty(B, __JEEFO_PROPERTY_НЭР__, __JEEFO_PROPERTY_ТОДОРХОЙЛОГЧ__); }());
    B.__jeefo_class__ = true; return B;}());
    const get_super = () => A;
    var C = (function () { var __JEEFO_СУПЕР_КЛАСС_0__ = get_super();  
        var __JEEFO_ӨМНӨХ_ХЭСЭГ__ = function (C) { C.prototype.method = function () {
            const method_name = "computed";
            __JEEFO_СУПЕР_КЛАСС_0__.prototype.method.call(this);
            __JEEFO_СУПЕР_КЛАСС_0__.prototype[method_name].call(this);
        }; };

        function C () {
            __JEEFO_СУПЕР_КЛАСС_0__.call(this, (function () {   function _C () { B.call(this); } _C.prototype = Object.create(B.prototype); _C.__jeefo_class__ = true; return _C;}()));
        } C.prototype = Object.create(__JEEFO_СУПЕР_КЛАСС_0__.prototype); __JEEFO_ӨМНӨХ_ХЭСЭГ__(C);
    C.__jeefo_class__ = true; return C;}());
    window.zzz = new C();
} catch (e) {
    const pre = document.createElement("pre");
    pre.appendChild(new Text(`${e}\n${e.stack}`));
    document.body.appendChild(pre);
}

async function main () {
    try {
        const root = new InvisibleComponent("(root)", {});
        await compile.from_elements([document.firstElementChild], root, false);
        await root.initialize();

        //console.log(root);
    } catch (e) {
        const pre = document.createElement("pre");
        pre.appendChild(new Text(`${e}\n${decodeURIComponent(e.stack)}`));
        document.body.appendChild(pre);
    }
}

window.addEventListener("load", main);
 });
