// @jeefo/component/directives/if.js
 jeefo.register("node_modules/@jeefo/component/directives/if.js", async (exports, module) => { const __dirname = "node_modules/@jeefo/component/directives", __filename = "node_modules/@jeefo/component/directives/if.js"; const require = path => { return jeefo.require(path, __filename, __dirname); }; /* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
* File Name   : if.js
* Created at  : 2017-09-17
* Updated at  : 2020-10-23
* Author      : jeefo
* Purpose     :
* Description :
* Reference   :
.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.*/
// ignore:start
"use strict";

/* globals*/
/* exported*/

// ignore:end

const jqlite               = (await require("@jeefo/jqlite"));
const Interpreter          = (await require("../interpreter"));
const ConditionalComponent = (await require("../components/conditional_component"));

const prop_comment   = Symbol("$comment");
const prop_component = Symbol("interpreter");

const compile_component = async (component, $comment) => {
    const wrapper = await new ConditionalComponent(
        "if--rendered", component.element, component
    );

    if (! wrapper.is_destroyed) {
        await wrapper.initialize();

        if (! wrapper.is_destroyed) {
            $comment.after(wrapper.$element);
            if (component.is_rendered) wrapper.trigger_render();
        }
    }
};

module.exports = {
    type     : "structure",
    selector : "if",
    priority : 900,

    controller : (function () { function IfDirective () {} 
        IfDirective.prototype.on_init = function ($element, component) {
            const { expression } = component;
            const comment = document.createComment(` If: ${ expression } `);

            component.interpreter = new Interpreter(expression, component);
            component.element.removeAttribute("if");

            this[prop_comment]   = jqlite(comment);
            this[prop_component] = component;

            $element.before(this[prop_comment]);
            $element.remove();

            return this.on_digest();
        };

        IfDirective.prototype.on_digest = async function () {
            const $comment  = this[prop_comment];
            const component = this[prop_component];

            const value = component.interpreter.get_value();

            if (value) {
                if (! component.children.length) {
                    await compile_component(component, $comment);
                }
            } else if (component.children.length) {
                await component.children[0].destroy();
            }
        };
    IfDirective.__jeefo_class__ = true; return IfDirective;}())
};
 });
