// @jeefo/material/components/card/index.js
 jeefo.register("node_modules/@jeefo/material/components/card/index.js", async (exports, module) => { const __dirname = "node_modules/@jeefo/material/components/card", __filename = "node_modules/@jeefo/material/components/card/index.js"; const require = path => { return jeefo.require(path, __filename, __dirname); }; /* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
* File Name   : index.js
* Created at  : 2019-09-30
* Updated at  : 2021-02-24
* Author      : jeefo
* Purpose     :
* Description :
* Reference   :
.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.*/
// ignore:start
"use strict";

/* globals*/
/* exported*/

// ignore:end

const Observer      = (await require("@jeefo/observer"));
const theme_service = (await require("../../services/theme"));

exports.selector = "md-card";

exports.style = (`.md-card{box-shadow:0px 2px 1px -1px rgba(0, 0, 0, 0.2),0px 1px 3px 0px rgba(0, 0, 0, 0.12),0px 1px 1px 0px rgba(0, 0, 0, 0.14);z-index:1;display:block;overflow:hidden;border-radius:4px}.md-card--outlined{box-shadow:0px 0px 0px 0px rgba(0, 0, 0, 0.2),0px 0px 0px 0px rgba(0, 0, 0, 0.12),0px 0px 0px 0px rgba(0, 0, 0, 0.14);z-index:0;border:1px solid;box-sizing:border-box}.md-card--shape--square{border-radius:unset}`
/* space filler
















*/);

theme_service.set_default({
    ".md-card--outlined": {
        "border-color": "$divider-color",
    },
});

theme_service.register_template((`.md-card--outlined{content:"props:[border-color]"}`
/* space filler





*/));

exports.template = `<md-surface><jf-content></jf-content></md-surface>`;

exports.bindings = {
    shape   : '@',
    variant : '@',
};

exports.controller = (function () { function MDCard () {} 
    MDCard.prototype.on_init = function ($element) {
        let variant, shape;
        const observer = new Observer(this);
        $element.add_class("md-card");

        const on_varient_change = value => {
            if (variant) $element.remove_class(`md-card--${variant}`);
            if (value) {
                variant = value.toLowerCase();
                $element.add_class(`md-card--${variant}`);
            }
        };

        const on_shape_change = value => {
            if (shape) $element.remove_class(`md-card--shape--${shape}`);
            if (value) {
                shape = value.toLowerCase();
                $element.add_class(`md-card--shape--${shape}`);
            }
        };

        on_shape_change(this.shape);
        on_varient_change(this.variant);
        observer.on("shape"   , on_shape_change);
        observer.on("variant" , on_varient_change);

    };
MDCard.__jeefo_class__ = true; return MDCard;}());

exports.controller_name = "$md_card";
 });
