// @jeefo/material/services/theme/index.js
 jeefo.register("node_modules/@jeefo/material/services/theme/index.js", async (exports, module) => { const __dirname = "node_modules/@jeefo/material/services/theme", __filename = "node_modules/@jeefo/material/services/theme/index.js"; const require = path => { return jeefo.require(path, __filename, __dirname); }; /* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
* File Name   : index.js
* Created at  : 2019-12-02
* Updated at  : 2021-02-27
* Author      : jeefo
* Purpose     :
* Description :
* Reference   :
.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.*/
// ignore:start
"use strict";

/* globals*/
/* exported*/

// ignore:end

const MDTheme = (await require("./theme"));

const parser = document.createElement("style");
parser.setAttribute("md-theme--parser", '');
document.head.appendChild(parser);

const parse_css_rules = css_text => {
    parser.textContent = css_text;

    const result     = [];
    const {sheet}    = parser;
    const {cssRules} = sheet;
    for (let i = 0; i < cssRules.length; i += 1) result.push(cssRules[i]);

    while (cssRules.length) sheet.deleteRule(0);
    parser.removeChild(parser.firstChild);

    return result;
};

var MDThemeService = (function () { 
    function MDThemeService () {
        this.themes = [];
    }

    MDThemeService.prototype.add_rules = function (css_text) {
        const {sheet} = this.style_element;

        for (const theme of this.themes) {
            const rules = parse_css_rules(css_text);
            for (const rule of rules) {
                const css_rule = `${theme.selector} ${rule.cssText}`;
                sheet.insertRule(css_rule, sheet.cssRules.length);
            }
        }
    };

    MDThemeService.prototype.set_default = function (defaults) {
        for (const theme of this.themes) {
            theme.default_sheet.add(theme, defaults);
        }
    };

    MDThemeService.prototype.register_theme = function ({name, colors, rules, sheet}) {
        this.themes.push(new MDTheme(name, colors, rules, sheet));
    };

    MDThemeService.prototype.register_template = function (css_text) {
        const css_rules = parse_css_rules(css_text);
        for (const theme of this.themes) {
            theme.eval_css_rules(css_rules);
        }
    };
MDThemeService.__jeefo_class__ = true; return MDThemeService;}());

module.exports = new MDThemeService();
 });
